deepinv/optim/phase_retrieval.py

```diff
@@ -129,3 +135,108 @@ def cosine_similarity(a: torch.Tensor, b: torch.Tensor):
+        \end{equation*}
+    
+    where :math:`B` is the linear operator of the phase retrieval class, :math:`T(\cdot)` is a preprocessing function for the measurements, and :math:`I` is the identity matrix of corresponding dimensions. Parameter :math:`\lambda` tunes the strength of regularization.
+
+    To find the principal eigenvector, the function runs power iteration which is given by
+
+    .. math::
+        \begin{equation*}
+        \begin{aligned}
+        x_{k+1} &= M x_k \\
+        x_{k+1} &= \frac{x_{k+1}}{\|x_{k+1}\|},
+        \end{aligned}
+        \end{equation*}
+  
+    :param torch.Tensor y: Measurements.
+    :param deepinv.physics physics: Instance of the physics modeling the forward matrix.
+    :param torch.Tensor x: Initial guess for the signals :math:`x_0`.
+    :param int n_iter: Number of iterations.
+    :param function preprocessing: Function to preprocess the measurements. Default is :math:`\max(1 - 1/x, -5)`.
+    :param float lamb: Regularization parameter. Default is 10.
+    :param bool log: Whether to log the metrics. Default is False.
+    :param function log_metric: Metric to log. Default is cosine similarity.
+    :param bool early_stop: Whether to early stop the iterations. Default is True.
+    :param float rtol: Relative tolerance for early stopping. Default is 1e-5.
+
+    :return: The estimated signals :math:`x`.
+    """
+    if x is None:
+        # always use randn for initial guess, never use rand!
+        x = torch.randn(
+            (y.shape[0],) + physics.input_shape,
+            dtype=physics.dtype,
+            device=physics.device,
+        )
+
+    if log is True:
+        metrics = []
+
+    #! estimate the norm of x using y
+    #! for the i.i.d. case, we have norm(x) = sqrt(sum(y)/A_squared_mean)
+    #! for the structured case, when the mean of the squared diagonal elements is 1, we have norm(x) = sqrt(sum(y)), otherwise y gets scaled by the mean to the power of number of layers
+    norm_x = torch.sqrt(y.sum())
+
+    x = x.to(torch.complex64)
+    # y should have mean 1
+    y = y / torch.mean(y)
+    diag_T = preprocessing(y, physics)
+    diag_T = diag_T.to(torch.complex64)
+    for i in range(n_iter):
+        x_new = physics.B(x)
+        x_new = diag_T * x_new
+        x_new = physics.B_adjoint(x_new)
+        x_new = x_new + lamb * x
+        x_new = x_new / torch.linalg.norm(x_new)
+        if log:
+            metrics.append(log_metric(x_new, x_true))
+        if early_stop:
+            if torch.linalg.norm(x_new - x) / torch.linalg.norm(x) < rtol:
+                print(f"Power iteration early stopped at iteration {i}.")
+                break
+        x = x_new
+    #! change the norm of x so that it matches the norm of true x
+    x = x * norm_x
+    if log:
+        return x, metrics
+    else:
+        return x
+
+
+def spectral_methods_wrapper(y, physics, n_iter=5000, **kwargs):
+    r"""
+    Wrapper for spectral methods.
```

Comment:
Add a little more details why this wrapper can be useful? e.g. initialization in deepinv optim